Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -308,16 +308,28 @@ namespace Tensile
 
         if(sizeMapping.persistentKernel != 0)
         {
-            size_t persistentGroups = dynamic_cast<AMDGPU const&>(hardware).computeUnitCount
-                                      * sizeMapping.persistentKernel;
+            size_t cuCount       = dynamic_cast<AMDGPU const&>(hardware).computeUnitCount;
+            size_t finalPKValue  = sizeMapping.persistentKernel;
             size_t problemGroups = rv.numWorkGroups.x * rv.numWorkGroups.y;
             if(sizeMapping.persistentKernelAlongBatch)
             {
                 problemGroups *= rv.numWorkGroups.z;
                 rv.numWorkGroups.z = 1;
             }
-            rv.numWorkGroups.x = std::min(persistentGroups, problemGroups);
-            rv.numWorkGroups.y = 1;
+
+            if(finalPKValue == -1)
+            {
+                // 1. Get the largest pk value (ex.3)
+                //    which can make the PK.G (ex.3*120=360) <= problemGroups (ex.433)
+                // 2. Scale by 5/8 (can try 0.5~1, to control the tiles-per-workgroup = 1~2)
+                finalPKValue = 5 * (problemGroups / cuCount) / 8;
+                finalPKValue = std::max(finalPKValue, (size_t)1);
+                //std::cout << "final persistent kernel value: " << finalPKValue << std::endl;
+            }
+
+            size_t persistentGroups = cuCount * finalPKValue;
```

Comment:
Repeated in ContractionProblem - these should probably be combined somehow since they involve a heuristic that looks like you might try to adjust in the future.

Reply:
Yes, I've thought about this when I did this part. 
However, the function "checkPersistentKernelEligibility" is not called int the path of rocBLAS side. 
(Also, we want to allow the tuned logic always runable in rocBLAS, since if it was tuned in Tensile, we have no reason to block it from running in rocBLAS). So it turns out I ro-do the PKn1-eval to get the final PKValue in case we wouldn't have done it (checkEligibility) in rocBLAS.
But I agree on make this part more elegant, maybe will re-design this in the next PR.